Add the missing contrib documentation

pwaller opened the pull request:
* Document a bit about what's in the contrib directory.
* Add a manyclangs-setup.sh which installs prerequisites.

Signed-off-by: Peter Waller <p@pwaller.net>

veselink1 commented on contrib/manyclangs-setup.sh:
```diff
@@ -0,0 +1,208 @@
+    then
+        if [ ! -d ccache ]
+        then
+            git clone --depth 1 --single-branch --branch v4.6.1 https://github.com/ccache/ccache
```
You mention that we use ccache v4.6.1 because it has some fixes in it. Any idea if those fixes are critical to building the packs? Because if they are, with our current detection strategy of `command -v ccache` we might not detect the case where an older ccache is already available in PATH.

pwaller replied:
Good point, I think I will unconditionally build ccache for now, and pin it. I don't have a specific thing I can easily point to, except for experience. I expect this would be unnecessary at some point the future.

veselink1 commented on contrib/manyclangs-setup.sh:
```diff
@@ -0,0 +1,208 @@
+    then
+        if [ ! -d jobclient ]
+        then
+            git clone https://github.com/olsner/jobclient
```
Is it a good idea to checkout a specific commit here, for security reasons?

pwaller replied:
Yes, let's do that.

veselink1 commented on contrib/README.md:
```diff
@@ -0,0 +1,70 @@
+# Contrib scripts
+
+This directory anything which isn't exactly elfshaker but is nonetheless useful.
```
? This directory ~~anything which~~ isn't exactly elfshaker but is nonetheless useful.

pwaller replied:
Rewritten:
> This directory contains scripts which make use of elfshaker, for example
scripts to build a [manyclangs](https://github.com/elfshaker/manyclangs) pack.

veselink1 commented on contrib/README.md:
```diff
@@ -0,0 +1,70 @@
+
+* To mitigate this, multiple builds run in parallel, but that conflicts with
+  making efficient use of incremental builds.
+* Therefore, builds need to be incremental, and it's necessary to multiple of
```
There might be something wrong with this sentence. Doesn't appear to make much sense.

veselink1 commented on contrib/README.md:
```diff
@@ -0,0 +1,70 @@
+  them run in parallel.
+  * Incremental builds start at a point in time and moves forward
+    commit-by-commit.
+    * This means a few tens-of-seconds build for a typical 'touch a few
```
I know what you mean here but I found it a bit hard to parse. Suggestion:

"The incremental builds are fast, in the order of tens of seconds per build, and they can be run in parallel reasonably well because most of the work done for each build is not heavily parallelized."

pwaller replied:
Rewritten much of the block:
>  * Incremental builds start at a point in time and move forward
    commit-by-commit.
>  * This results in a few tens-of-seconds build for a typical 'touch a few
    translation units' commit. However, such a build will only consume a few
    CPUs because they are bottlenecked on a small number (much fewer than
    `$NCPU`) of large translation units for that time.
>  * Saturating the machine to use all of the available resource is achieved by
    building from multiple points in time simultaneously.
>  * This introduces a problem of build contention, running more compiler
    procesess in parallel than necessary is wasteful and risks deadlocking the
    machine for example via an out-of-memory condition. To mitigate this, a
    jobserver-aware ninja is used, which ensure that only as many compilers run
    concurrently as there are available CPU cores.

veselink1 reviewed (approved):
LGTM with typos fixed and minor Qs answered, thanks. Let me know if you want me to fix any of these.

pwaller commented:
Thanks for spotting my mistakes, that was very helpful. I hope I've addressed your comments without introducing further mistakes. I've done another test of the setup script since my changes and all seems to work.

pwaller commented:
Btw your comments remind me of John Reghr's blog post [You Might as Well Be a Great Copy Editor](https://blog.regehr.org/archives/1471). It's eye opening for me to see how many little mistakes crept in, although I'm not too surprised since I think I wrote the whole thing in about five minutes and didn't do much editing, so thanks for spending the time to help me iterate.

veselink1 commented:
Changes look good, thanks.
